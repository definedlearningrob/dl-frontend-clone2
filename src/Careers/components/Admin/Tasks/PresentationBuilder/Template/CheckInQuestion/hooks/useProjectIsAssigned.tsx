import { useMemo } from 'react';
import { useParams } from 'react-router-dom';
import { useQuery } from '@apollo/client';
import { isEmpty } from 'lodash-es';

import teamProjectQuery from '@pbl/graphql/student/queries/teamProject';
import projectQuery from '@pbl/graphql/student/queries/project';

import { useRole } from '@shared/hooks/useRole';

export const useProjectIsAssigned = () => {
  const { isStudent } = useRole();
  const { projectId, teamId } = useParams<{ projectId: string; teamId: string }>();

  const properQuery = useMemo(() => (teamId ? teamProjectQuery : projectQuery), [teamId]);

  const queryVariables = useMemo(
    () =>
      teamId
        ? { id: projectId, teamId, track: true, trackPresentation: true }
        : { id: projectId, track: true, trackPresentation: true },
    [teamId]
  );

  const { data, error, loading } = useQuery(properQuery, {
    variables: queryVariables,
    skip: !isStudent,
  });

  if (!isStudent || loading || error) {
    return false;
  }

  return !isEmpty(data?.project?.assignedAt);
};
