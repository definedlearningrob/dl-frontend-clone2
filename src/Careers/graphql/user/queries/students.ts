import { gql } from '@apollo/client';

export default gql`
  query Students($page: Int, $perPage: Int, $filter: StudentFilter, $scope: ArchivableStatus) {
    students(page: $page, perPage: $perPage, filter: $filter, scope: $scope) {
      nodes {
        uuid
        firstName
        lastName
        entity {
          uuid
          name
        }
      }
      nodesCount
      pagesCount
    }
  }
`;
