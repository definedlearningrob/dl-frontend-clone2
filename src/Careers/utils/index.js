const isObject = (object) => object !== null && typeof object === 'object';

export const isEqual = (object1, object2) => {
  if (object1 === object2) return true;

  if (
    object1 === null ||
    object2 === null ||
    typeof object1 !== 'object' ||
    typeof object2 !== 'object'
  )
    return false;

  const keys1 = Object.keys(object1);
  const keys2 = Object.keys(object2);

  if (keys1.length !== keys2.length) {
    return false;
  }

  for (let key of keys1) {
    const val1 = object1[key];
    const val2 = object2[key];
    const areObjects = isObject(val1) && isObject(val2);
    if (
      // eslint-disable-next-line
      areObjects && !isEqual(val1, val2) ||
      // eslint-disable-next-line
      !areObjects && val1 !== val2
    ) {
      return false;
    }
  }

  return true;
};

export const getByStringKey = (object, key) => {
  const keys = key.split('.');

  return keys.reduce((acc = {}, key) => acc[key], object);
};
