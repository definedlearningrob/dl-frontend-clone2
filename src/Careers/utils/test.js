/* eslint-disable react/prop-types */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { createMemoryHistory } from 'history';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { Router, Route } from 'react-router-dom';

import i18n from '@dc/i18n';

export function renderWithRouter(
  ui,
  { route = '/', routePath = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  const Wrapper = ({ children }) => (
    <Router history={history}>
      <I18nextProvider i18n={i18n}>
        <Route path={routePath} render={() => children} />
      </I18nextProvider>
    </Router>
  );

  return {
    ...render(ui, { wrapper: Wrapper }),
    history,
  };
}

export function renderWithReduxProvider(
  ui,
  {
    initialState = {
      session: {
        loginError: {},
        user: null,
      },
    },
    store = configureMockStore([thunk])(initialState),
  } = {}
) {
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </Provider>
  );

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
}

export function renderWithRouterAndReduxProvider(
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    initialState = {
      session: {
        loginError: {},
        user: {},
      },
    },
    store = configureMockStore([thunk])(initialState),
  } = {},
  ...options
) {
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <Router history={history}>
        <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
      </Router>
    </Provider>
  );

  return {
    ...render(ui, { wrapper: Wrapper, ...options }),
    history,
  };
}

export function renderWithI18N(ui) {
  const Wrapper = ({ children }) => <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;

  return {
    ...render(ui, { wrapper: Wrapper }),
  };
}
