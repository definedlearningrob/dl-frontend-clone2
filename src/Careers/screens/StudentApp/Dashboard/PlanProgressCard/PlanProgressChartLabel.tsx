import { LabelProps } from 'recharts';
import { PolarViewBox } from 'recharts/types/util/types';

import { useBreakpointUp } from '@shared/hooks/useBreakpointUp';

export const PlanProgressChartLabel = ({ viewBox, value }: LabelProps) => {
  const isFullHD = useBreakpointUp({ breakpoint: 'xxxl' });

  const { cx, cy } = viewBox as PolarViewBox;

  return (
    <g>
      <text
        dominantBaseline='middle'
        fill='#FCFCFC'
        fontSize={isFullHD ? 18 : 14}
        fontWeight={700}
        textAnchor='middle'
        x={cx}
        y={cy}>
        {value}
      </text>
    </g>
  );
};
