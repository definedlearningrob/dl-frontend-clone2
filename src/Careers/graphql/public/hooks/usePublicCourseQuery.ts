import { useQuery } from '@apollo/client';

import { PUBLIC_COURSE_QUERY } from '@dc/graphql/public/queries/course';

type Params = {
  skip?: boolean;
  code: string;
  shareId: string;
};

export const usePublicCourseQuery = ({ skip, code, shareId }: Params) =>
  useQuery(PUBLIC_COURSE_QUERY, {
    skip,
    variables: { code, shareId, track: true },
  });
