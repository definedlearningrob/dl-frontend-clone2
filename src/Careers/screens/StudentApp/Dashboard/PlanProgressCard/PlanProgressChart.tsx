import { Cell, Label, Pie, PieChart } from 'recharts';

import { useBreakpointUp } from '@shared/hooks/useBreakpointUp';
import { convertToPercentage } from '@shared/utils/convertToPercentage';

import { PlanProgressChartLabel } from './PlanProgressChartLabel';

type Props = {
  value: number;
};

export const PlanProgressChart = ({ value }: Props) => {
  const isFullHD = useBreakpointUp({ breakpoint: 'xxxl' });

  const chartSize = isFullHD ? 164 : 100;
  const chartLabel = convertToPercentage(value);
  const chartData = [
    { value, fill: '#ED7700' },
    { value: 1 - value, fill: '#FCFCFC' },
  ];

  const isInProgress = value > 0 && value < 1;

  return (
    <PieChart
      height={chartSize}
      margin={{ bottom: 0, top: 0, left: 0, right: 0 }}
      width={chartSize}>
      <Pie
        data={chartData}
        dataKey='value'
        endAngle={-270}
        innerRadius='65%'
        isAnimationActive={false}
        outerRadius='100%'
        paddingAngle={isInProgress ? 4 : 0}
        startAngle={90}
        stroke='transparent'>
        {chartData.map((entry, index) => (
          <Cell key={`cell-${index}`} fill={entry.fill} />
        ))}
        <Label content={<PlanProgressChartLabel value={chartLabel} />} position='center' />
      </Pie>
    </PieChart>
  );
};
