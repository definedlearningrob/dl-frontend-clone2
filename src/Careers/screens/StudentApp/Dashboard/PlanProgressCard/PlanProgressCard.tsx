import { useTranslation } from 'react-i18next';
import { usePlanProgressQuery } from '@graphql/dc/students/hooks';
import { isEmpty } from 'lodash-es';
import { useRef } from 'react';

import Card from '@shared/components/Card/Card';
import { Kicker } from '@shared/components/Kicker';
import { useBreakpointUp } from '@shared/hooks/useBreakpointUp';
import Link from '@shared/components/Link';
import SkeletonRectangle from '@shared/components/Skeleton/Rectangle/Rectangle';
import { useIsTruncated } from '@shared/hooks/useIsTruncated';
import { Tooltip } from '@shared/components/Tooltip';

import { PlanProgressChart } from './PlanProgressChart';

export const PlanProgressCard = () => {
  const { t } = useTranslation();
  const isFullHD = useBreakpointUp({ breakpoint: 'xxxl' });
  const planNameRef = useRef<HTMLHeadingElement>(null);
  const isPlanNameTruncated = useIsTruncated({ ref: planNameRef });

  const { data, loading } = usePlanProgressQuery();

  if (loading) {
    return <SkeletonRectangle className='!h-[224px] xxxl:!h-[292px]' radius='sm' />;
  }

  if (isEmpty(data?.plans)) {
    return null;
  }

  const [plan] = data!.plans;
  const {
    progress: { completed, total },
  } = plan!;

  return (
    <Card className='bg-gradient-primary-600 flex flex-col gap-base !text-white shadow-300 xl:justify-normal justify-evenly'>
      <div className='flex items-center justify-between gap-sm xxxl:gap-base'>
        <div>
          <Kicker className='!text-neutral-300 !mb-sm xxxl:!mb-base' size={isFullHD ? 'md' : 'sm'}>
            {t('student.dashboard.yourPlanProgress')}
          </Kicker>
          <Tooltip delayDuration={500} disabled={!isPlanNameTruncated} message={plan?.name}>
            <h5 ref={planNameRef} className='line-clamp-2 text-sm xxxl:text-base mb-xs xxxl:mb-sm'>
              {plan?.name}
            </h5>
          </Tooltip>
          <span className='text-neutral-300 leading-lg text-xxs xxxl:text-xs'>
            {t('student.dashboard.trackPlanProgress')}
          </span>
        </div>
        <div className='shrink-0'>
          <PlanProgressChart value={completed / (total || 1)} />
        </div>
      </div>
      <Link
        className='!text-white hover:!bg-white hover:!text-primary-500 w-full'
        size={isFullHD ? 'lg' : 'md'}
        to='/plans'
        variant='primary-outlined'>
        {t('student.dashboard.viewPlan')}
      </Link>
    </Card>
  );
};
