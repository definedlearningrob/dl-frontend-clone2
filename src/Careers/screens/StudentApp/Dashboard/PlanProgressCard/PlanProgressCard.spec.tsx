import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { PlanProgressQuery } from '@graphql/dc/students/operations';
import { PlanProgressDocument } from '@graphql/dc/students/hooks';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithRouter } from '@shared/utils/test';

import { PlanProgressCard } from './PlanProgressCard';

const planProgressMock: MockedResponse<PlanProgressQuery> = {
  request: {
    query: PlanProgressDocument,
  },
  result: {
    data: {
      plans: [
        {
          id: '1',
          name: 'Defined Portrait of a Graduate',
          progress: {
            completed: 7,
            total: 36,
          },
        },
      ],
    },
  },
};

const renderComponent = () =>
  renderWithRouter(
    <MockedProvider mocks={[planProgressMock]}>
      <PlanProgressCard />
    </MockedProvider>
  );

describe('PlanProgressCard', () => {
  it('should render correctly', async () => {
    const { container } = renderComponent();

    expect(
      await screen.findByRole('heading', { name: 'Defined Portrait of a Graduate' })
    ).toBeInTheDocument();
    expect(screen.getByText('19%')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should navigate to plans after clicking on the button', async () => {
    const { history } = renderComponent();
    history.push = jest.fn();

    const button = await screen.findByRole('button', { name: 'View plan' });

    userEvent.click(button);

    expect(history.push).toHaveBeenCalledWith('/plans');
  });
});
