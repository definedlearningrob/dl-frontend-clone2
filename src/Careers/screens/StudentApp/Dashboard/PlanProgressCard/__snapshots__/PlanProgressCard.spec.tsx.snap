// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanProgressCard should render correctly 1`] = `
<div>
  <div
    class="card bg-gradient-primary-600 flex flex-col gap-base !text-white shadow-300 xl:justify-normal justify-evenly"
  >
    <div
      class="flex items-center justify-between gap-sm xxxl:gap-base"
    >
      <div>
        <div
          class="kicker sm default !text-neutral-300 !mb-sm xxxl:!mb-base"
        >
          Your plan progress
        </div>
        <span
          data-state="closed"
        >
          <h5
            class="line-clamp-2 text-sm xxxl:text-base mb-xs xxxl:mb-sm"
          >
            Defined Portrait of a Graduate
          </h5>
        </span>
        <span
          class="text-neutral-300 leading-lg text-xxs xxxl:text-xs"
        >
          Track your personal progress within the plan
        </span>
      </div>
      <div
        class="shrink-0"
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 100px; height: 100px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="100"
            style="width: 100%; height: 100%;"
            viewBox="0 0 100 100"
            width="100"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="100"
                  width="100"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer recharts-pie-sector"
                tabindex="-1"
              >
                <path
                  class="recharts-sector"
                  cx="50"
                  cy="50"
                  d="M 50,0
    A 50,50,0,
    0,1,
    96.50308805206615,31.62983936864509
  L 80.227007233843,38.05939558961931
            A 32.5,32.5,0,
            0,0,
            50,17.5 Z"
                  fill="#ED7700"
                  name="0"
                  role="img"
                  stroke="transparent"
                  tabindex="-1"
                />
              </g>
              <g
                class="recharts-layer recharts-pie-sector"
                tabindex="-1"
              >
                <path
                  class="recharts-sector"
                  cx="50"
                  cy="50"
                  d="M 97.67124649456483,34.91847959738673
    A 50,50,0,
    1,1,
    46.51217631279372,0.12179748700879145
  L 47.73291460331592,17.579168366555713
            A 32.5,32.5,0,
            1,0,
            80.98631022146714,40.19701173830137 Z"
                  fill="#FCFCFC"
                  name="1"
                  role="img"
                  stroke="transparent"
                  tabindex="-1"
                />
              </g>
              <g>
                <text
                  dominant-baseline="middle"
                  fill="#FCFCFC"
                  font-size="14"
                  font-weight="700"
                  text-anchor="middle"
                  x="50"
                  y="50"
                >
                  19%
                </text>
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
    <a
      href="/plans"
    >
      <button
        class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] !text-white hover:!bg-white hover:!text-primary-500 w-full -primary-outlined text-primary-500 hover:bg-primary-200 py-xs px-base leading-[22px] border"
        data-testid="button"
        type="button"
      >
        <span
          class="items-center flex justify-center gap-xs"
        >
          View plan
        </span>
      </button>
    </a>
  </div>
</div>
`;
